### PR TITLE
Complete action plan 

### PR DESCRIPTION
**Pull request by stevomcallister**
New Endpoint to provide complete action plan activities. Endpoint takes a goal reference and a note. 

Endpoint validates that the goal is in the correct state to be completed - 
	-	status is ACTIVE
	-  all child steps are COMPLETE
	
Other downstream activities are completed, timeline entry created and telemetry event.

___

While writing this I found that the test suite ran fine locally and failed almost 100% of the time in CI. As a result most of the commit history is me trying to find out why. The issue seems to be due to the async creation of events. When the time line tests run the timeline entries are not present and fail the test. I wondered for quite some time if it was due to using the same prisoner id multiple times and hence the additional events created by completing a goal were breaking the tests.

Turns out that the time line tests failed some of the time in builds prior to this pr - hence the additional async calls were the straw that broke this camels back. 

TL;DR - I increased the await timeout in the the timeline test.

### CODE REVIEW

File: domain/learningandworkprogress/src/testFixtures/kotlin/uk/gov/justice/digital/hmpps/domain/CommonBuilders.kt [hunk 1]
```diff
@@ -11,4 +11,17 @@ fun aValidPrisonNumber() = "A1234BC"
 
 fun anotherValidPrisonNumber() = "B5678CD"
 
+fun randomValidPrisonNumber(): String {
```
Comment (stevomcallister) [inline domain/learningandworkprogress/src/testFixtures/kotlin/uk/gov/justice/digital/hmpps/domain/CommonBuilders.kt:14]: I created this while barking up the wrong tree - thought that I was getting conflicts due to reusing the same prison number, this was not the case but this function may be handy anyway so left it in.

File: domain/personallearningplan/src/main/kotlin/uk/gov/justice/digital/hmpps/domain/personallearningplan/Exceptions.kt [hunk 2]
```diff
@@ -40,6 +41,9 @@ class InvalidGoalStateException(
 ) :
   RuntimeException("Could not ${action.name.lowercase()} goal with reference [$goalReference] for prisoner [$prisonNumber]: Goal was in state [$status] that can't be ${action.name.lowercase()}d")
 
+class ActiveGoalStepsException(val goalReference: UUID, val prisonNumber: String) :
```
Comment (stevomcallister) [inline domain/personallearningplan/src/main/kotlin/uk/gov/justice/digital/hmpps/domain/personallearningplan/Exceptions.kt:44]: Maybe a poor name choice but couldn't think of anything better - Exception to be raised when an attempt is made to complete a goal but it has active child steps.

File: src/integrationTest/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/resource/GetTimelineTest.kt [hunk 3]
```diff
@@ -73,7 +74,7 @@ class GetTimelineTest : IntegrationTestBase() {
   @Test
   fun `should not get timeline given prisoner has no timeline`() {
     // Given
-    val prisonNumber = aValidPrisonNumber()
```
Comment (stevomcallister) [inline src/integrationTest/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/resource/GetTimelineTest.kt:76]: As mentioned earlier, not necessary in the end but not a problem either :-/

File: src/integrationTest/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/resource/CompleteGoalTest.kt [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+      prisonNumber = prisonNumber,
+      createGoalRequest = createGoalRequest,
+    )
+    val actionPlan = getActionPlan(prisonNumber)
+    val goal = actionPlan.goals[0]
+    val goalReference = goal.goalReference
+    return goalReference
+  }
+
+  fun completeSteps(prisonNumber: String) {
```
Comment (stevomcallister) [inline src/integrationTest/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/resource/CompleteGoalTest.kt:323]: Added this method to complete the child goal steps - may need this elsewhere at some point so is a candidate for adding to the super class ?.

File: src/main/resources/static/openapi/EducationAndWorkPlanAPI.yml [hunk 5]
```diff
@@ -334,6 +334,33 @@ paths:
...
+      - $ref: "#/components/parameters/prisonNumberPathParameter"
+      - $ref: "#/components/parameters/goalReferencePathParameter"
+    put:
+      summary: Completes a Goal
+      description: |
+        Changes a goals status to 'COMPLETE' and records the completion note.
+
+        **Role Requirements:**
+        Access to this endpoint requires one of the following roles:
+          - `ROLE_EDUCATION_AND_WORK_PLAN__GOALS__RO` (Read-Only)
```
Comment (nathanrussell-moj-digital) [inline src/main/resources/static/openapi/EducationAndWorkPlanAPI.yml:348]: Not the RO role? 🤔

File: src/integrationTest/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/resource/UpdateGoalTest.kt [hunk 6]
```diff
@@ -32,7 +32,7 @@ import uk.gov.justice.digital.hmpps.educationandworkplanapi.withBody
 class UpdateGoalTest : IntegrationTestBase() {
 
   companion object {
-    private const val URI_TEMPLATE = "/action-plans/{prisonNumber}/goals/{goalReference}"
+    const val URI_TEMPLATE = "/action-plans/{prisonNumber}/goals/{goalReference}"
```
Comment (nathanrussell-moj-digital) [inline src/integrationTest/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/resource/UpdateGoalTest.kt:35]: Not that I massively care 🤣 , but is there a change somewhere that I've missed that uses this? Or is it a change as a result of some earlier work?

File: src/main/resources/static/openapi/EducationAndWorkPlanAPI.yml (hunk 5 above)
Comment (stevomcallister) [reply src/main/resources/static/openapi/EducationAndWorkPlanAPI.yml:348]: copy paste from the archive/unarchive - which are also wrong 🙄

File: src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/JpaGoalPersistenceAdapter.kt [hunk 7]
```diff
@@ -92,6 +94,17 @@ class JpaGoalPersistenceAdapter(
     )
   }
 
+  override fun completeGoal(prisonNumber: String, completeGoalDto: CompleteGoalDto): Goal? {
+    return getGoalEntityByPrisonNumberAndGoalReference(prisonNumber, completeGoalDto.reference)?.let { goalEntity ->
+      goalEntity.apply {
+        status = GoalStatus.COMPLETED
+      }
+      goalEntity.steps().forEach { it.apply { it.status = StepStatus.COMPLETE } }
```
Comment (stevomcallister) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/JpaGoalPersistenceAdapter.kt:102]: Also complete child steps

File: src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/service/AsyncGoalEventService.kt [hunk 8]
```diff
@@ -102,6 +113,23 @@ class AsyncGoalEventService(
     )
   }
 
+  private fun recordGoalCompletedTimelineEvent(
+    prisonNumber: String,
+    previousGoal: Goal,
+    updatedGoal: Goal,
+    correlationId: UUID = UUID.randomUUID(),
+  ) {
+    val timelineEvents = timelineEventFactory.getStepUpdatedEvents(
```
Comment (stevomcallister) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/service/AsyncGoalEventService.kt:122]: use existing code to work out which steps have changed and save all events

File: src/integrationTest/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/resource/UpdateGoalTest.kt (hunk 6 above)
Comment (stevomcallister) [reply src/integrationTest/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/resource/UpdateGoalTest.kt:35]: yes this can be privatised.

**Review by nathanrussell-moj-digital — APPROVED** [review]
LGTM 👍 Nice
